### PR TITLE
[WIP] [MXNET-325] Model parallelism tutorial.

### PR DESCRIPTION
**Pull request by indhub**
## Description ##

I'm working on a tutorial to show how model parallelism can be done in Gluon. The tutorial builds a stacked LSTM with different LSTM layers on different GPUs (like described in [this](https://mxnet.incubator.apache.org/faq/model_parallel_lstm.html) tutorial). The code borrows from [this](http://gluon.mxnet.io/chapter05_recurrent-neural-networks/rnns-gluon.html) gluon tutorial and makes it model parallel.

This is a Work In Progress. Before I write a tutorial around this code, I wanted to get the approach reviewed to make sure this is the correct way to do model parallelism in Gluon.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)

### CODE REVIEW

**Comment by indhub** [comment]
@safrooze @Ishitori @thomelane @ThomasDelteil

File: docs/tutorials/gluon/model_parallelism.md [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+
+
+```python
+test_L = eval(test_data)
+print('Best test loss %.2f, test perplexity %.2f'%(test_L, math.exp(test_L)))
+```
+
+    Best test loss 5.61, test perplexity 273.09
+
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: you are missing the download notebook place holder

File: docs/tutorials/gluon/model_parallelism.md [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+# Model parallelism
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: you haven't updated the `tutorials/index.md` 
please also update the nightly/tests/test_tutorial_config.txt` and run `python test_tutorial.py --tutorial gluon_model_parallelism` to make sure your notebook runs

File: docs/tutorials/gluon/model_parallelism.md [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+                words = line.split() + ['<eos>']
+                for word in words:
+                    ids[token] = self.dictionary.word2idx[word]
+                    token += 1
+
+        return mx.nd.array(ids, dtype='int32')
+```
+
+`MultiGPULSTM` creates stacked LSTM with layers spread across multiple GPUs. 
+For example, `MultiGPULSTM(0, [1, 2, 2, 1], 400, 200, 0.5)` will create a stacked LSTM with one layer on GPU(0), two layers on GPU(1), two layers on GPU(2), one layer on GPU(3) with a hidden size of 400 embedding size of 200 and dropout probability of .5.
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: `with a hidden size of 400 embedding size of 200`  -> `with a hidden size of 400 and embedding size of 200 `

File: docs/tutorials/gluon/model_parallelism.md [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    
+    def __init__(self, start_device, num_layers_list, num_hidden, input_size, dropout):
+        """Create a MultiGPULSTM. num_layers_list dictates how many layers of LSTM
+        gets places in which device. For example, [1, 2, 2, 1] will create a stacked LSTM
+        with one layer on GPU(0), two layers on GPU(1), two layers on GPU(2), one layer on GPU(3)"""
+        self.lstm_dict = collections.OrderedDict()
+        device_index = start_device
+        self.trainers = []
+        
+        for num_layers in num_layers_list:
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: maybe add a check here that makes sure `len(num_layers_list)` isn't bigger than what list_gpus() returns

File: docs/tutorials/gluon/model_parallelism.md [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    
+    def reset_optimizer(self, optimizer, optimizer_params=None):
+        self.encoder_trainer._init_optimizer(optimizer, optimizer_params)
+        self.decoder_trainer._init_optimizer(optimizer, optimizer_params)
+        self.lstm.reset_optimizer(optimizer, optimizer_params)
+```
+
+
+```python
+args_data = 'data/ptb.'
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: please use the `os.path.join('data','ptb.')`  formulation to make sure the path is understood on windows

File: docs/tutorials/gluon/model_parallelism.md [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+            epoch + 1, time.time() - start_time, val_L, math.exp(val_L)))
+
+```
+
+
+```python
+train()
+```
+
+    [Epoch 1 Batch 100] loss 7.20, perplexity 1340.10
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: you can replace this with, that will allow your generated notebook to not have all this output:

```
`[Epoch 1 Batch 100] loss 7.20, perplexity 1340.10` <!--notebook-skip-line-->

`[Epoch 1 Batch 200] loss 6.66, perplexity 781.10` <!--notebook-skip-line-->

`...` <!--notebook-skip-line-->

`[Epoch 1 Batch 4800] loss 5.59, perplexity 267.57` <!--notebook-skip-line-->

`[Epoch 1] time cost 267.24s, validation loss 5.63, validation perplexity 279.41` <!--notebook-skip-line-->
```

File: docs/tutorials/gluon/model_parallelism.md [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+    [Epoch 1] time cost 267.24s, validation loss 5.63, validation perplexity 279.41
+
+
+
+```python
+test_L = eval(test_data)
+print('Best test loss %.2f, test perplexity %.2f'%(test_L, math.exp(test_L)))
+```
+
+    Best test loss 5.61, test perplexity 273.09
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: replace with
```
`Best test loss 5.61, test perplexity 273.09`<!--notebook-skip-line-->
```

File: docs/tutorials/gluon/model_parallelism.md [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    if isinstance(hidden, (tuple, list)):
+        hidden = [detach(i) for i in hidden]
+    else:
+        hidden = hidden.detach()
+    return hidden
+```
+
+
+```python
+def eval(data_source):
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/model_parallelism.md]: `eval` is a poorly chose name IMO as there is a `eval` function in python.

**Comment by indhub** [comment]
@ThomasDelteil Thanks for the review. Like mentioned in the description, I haven't 'tutorialized' the code yet. I created a PR with just the code so that people can review the code to make sure I'm doing model parallelism right. If there is no objection for the approach I have taken, I'll add nice description, images and explain the code.

File: docs/tutorials/gluon/model_parallelism.md (hunk 2 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Thanks. I'll do this.

File: docs/tutorials/gluon/model_parallelism.md (hunk 1 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Will add.

File: docs/tutorials/gluon/model_parallelism.md (hunk 4 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Sure

File: docs/tutorials/gluon/model_parallelism.md (hunk 5 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Sure

File: docs/tutorials/gluon/model_parallelism.md (hunk 6 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Thanks for the tip.

File: docs/tutorials/gluon/model_parallelism.md (hunk 8 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Will rename.

File: docs/tutorials/gluon/model_parallelism.md (hunk 7 above)
Comment (indhub) [reply docs/tutorials/gluon/model_parallelism.md]: Sure

**Comment by indhub** [comment]
@mli

**Comment by indhub** [comment]
@ThomasDelteil When I run it on two GPUs, I see 100% utilization of both GPUs. There are no blocking operations in the code.
